### PR TITLE
feat: node fs module

### PR DESCRIPTION
**Pull request by Puellaquae**
* https://github.com/WasmEdge/WasmEdge/issues/1834
- [x] FileHandle
- [x] fsPromises.access
- [x] fsPromises.appendFile
- [x] <del>fsPromises.chmod</del>  **Unsupported**
- [x] <del>fsPromises.chown</del>  **Unsupported**
- [x] fsPromises.copyFile
- [x] fsPromises.cp
- [x] <del>fsPromises.lchmod</del>  **Unsupported**
- [x] <del>fsPromises.lchown</del>  **Unsupported**
- [x] fsPromises.lutimes
- [x] fsPromises.link
- [x] fsPromises.lstat
- [x] fsPromises.mkdir
- [x] fsPromises.mkdtemp
- [x] fsPromises.open
- [x] fsPromises.opendir
- [x] fsPromises.readdir
- [x] fsPromises.readFile
- [x] fsPromises.readlink
- [x] fsPromises.realpath
- [x] fsPromises.rename
- [x] fsPromises.rmdir
- [x] fsPromises.rm
- [x] fsPromises.stat
- [x] fsPromises.symlink
- [x] fsPromises.truncate
- [x] fsPromises.unlink
- [x] fsPromises.utimes
- [x] fsPromises.watch
- [x] fsPromises.writeFile
- [x] fsPromises.constants
- [x] fs.access
- [x] fs.accessSync
- [x] fs.appendFile
- [x] fs.appendFileSync
- [x] <del>fs.chmod</del>  **Unsupported**
- [x] <del>fs.chmodSync</del>  **Unsupported**
- [x] <del>fs.chown</del> **Unsupported**
- [x] <del>fs.chownSync</del>  **Unsupported**
- [x] fs.close
- [x] fs.closeSync
- [x] fs.copyFile
- [x] fs.copyFileSync
- [x] fs.cp
- [x] fs.cpSync
- [x] fs.createReadStream
- [x] fs.createWriteStream
- [x] fs.exists
- [x] fs.existsSync
- [x] <del>fs.fchmod</del>   **Unsupported**
- [x] <del>fs.fchmodSync</del>  **Unsupported**
- [x] <del>fs.fchown</del>   **Unsupported**
- [x] <del>fs.fchownSync</del>  **Unsupported**
- [x] fs.fdatasync
- [x] fs.fdatasyncSync
- [x] fs.fstat
- [x] fs.fstatSync
- [x] fs.fsync
- [x] fs.fsyncSync
- [x] fs.ftruncate
- [x] fs.ftruncateSync
- [x] fs.futimes
- [x] fs.futimesSync
- [x] <del>fs.lchmod</del>   **Unsupported**
- [x] <del>fs.lchmodSync</del>   **Unsupported**
- [x] <del>fs.lchown</del>   **Unsupported**
- [x] <del>fs.lchownSync</del>   **Unsupported**
- [x] fs.link
- [x] fs.linkSync
- [x] fs.lstat
- [x] fs.lstatSync
- [x] fs.lutimes
- [x] fs.lutimesSync
- [x] fs.mkdir
- [x] fs.mkdirSync
- [x] fs.mkdtemp
- [x] fs.mkdtempSync
- [x] fs.open
- [x] fs.openSync
- [x] fs.opendir
- [x] fs.opendirSync
- [x] fs.read
- [x] fs.readSync
- [x] fs.readdir
- [x] fs.readdirSync
- [x] fs.readFile
- [x] fs.readFileSync
- [x] fs.readlink
- [x] fs.readlinkSync
- [x] fs.readv
- [x] fs.readvSync
- [x] fs.realpath
- [x] fs.realpathSync
- [x] fs.rename
- [x] fs.renameSync
- [x] fs.rmdir
- [x] fs.rmdirSync
- [x] fs.rm
- [x] fs.rmSync
- [x] fs.stat
- [x] fs.statSync
- [x] fs.symlink
- [x] fs.symlinkSync
- [x] fs.truncate
- [x] fs.truncateSync
- [x] fs.unlink
- [x] fs.unlinkSync
- [x] <del>fs.unwatchFile</del> **Unsupported**
- [x] fs.utimes
- [x] fs.utimesSync
- [x] <del>fs.watch</del> **Unsupported**
- [x] <del>fs.watchFile</del> **Unsupported**
- [x] fs.write
- [x] fs.writeSync
- [x] fs.writeFile
- [x] fs.writeFileSync
- [x] fs.writev
- [x] fs.writevSync
- [x] fs.Dir
- [x] fs.Dirent
- [ ] <del>fs.FSWatcher</del> **Unsupported**
- [ ] <del>fs.StatWatcher</del> **Unsupported**
- [x] fs.ReadStream
- [x] fs.Stats
- [x] fs.WriteStream
- [x] fs.constants

### CODE REVIEW

**Comment by juntao** [comment]
@Puellaquae Does nodejs have a CI test suite we can run to validate those API implementations? Thanks!

**Comment by Puellaquae** [comment]
> @Puellaquae Does nodejs have a CI test suite we can run to validate those API implementations? Thanks!

There seems no test suite for node lib api in nodejs, while deno has some test suite for its node api implementations.

**Comment by juntao** [comment]
Is it possible for us to run the deno tests then? Thanks.

**Comment by Puellaquae** [comment]
It need some works before running.

**Comment by juntao** [comment]
BTW, if you incorporated any Deno code (MIT license), including test cases, you should add a line of acknowledgment as a comment at the top of the file. Thanks

File: Cargo.toml [hunk 1]
```diff
@@ -22,7 +22,7 @@ libc = "0.2"
 url = "2.2.2"
 lazy_static = "1.4"
 encoding = "0.2"
-
+wasi = "0.11.0+wasi-snapshot-preview1"
```
Comment (L-jasmine) [inline Cargo.toml:25]: please select your need function to `src/event_loop/wasi_fs.rs` , and remove `wasi` use.

File: src/internal_module/fs.rs [hunk 2]
```diff
@@ -0,0 +1,862 @@
+use crate::event_loop::PollResult;
+use crate::quickjs_sys::*;
+use std::convert::TryInto;
+use std::fs;
+use std::fs::Permissions;
+use std::io;
+use std::os::wasi::fs::{FileTypeExt, MetadataExt};
```
Comment (L-jasmine) [inline src/internal_module/fs.rs:7]: this feature need nigthly rust, but your can implement it use `std::os::wasi::prelude::FromRawFd` and `wasi_snapshot_preview1.fd_filestat_get`

File: src/internal_module/fs.rs [hunk 3]
```diff
@@ -579,45 +586,46 @@ fn open_sync(ctx: &mut Context, _this_val: JsValue, arg: &[JsValue]) -> JsValue
...
-                    | wasi::RIGHTS_FD_READ
-                    | wasi::RIGHTS_FD_READDIR;
-                let res = unsafe { wasi::path_open(3, 0, path.as_str(), oflag, right, 0, fdflag) };
+                } | wasi_fs::RIGHTS_FD_FILESTAT_GET
+                    | wasi_fs::RIGHTS_FD_SEEK
+                    | wasi_fs::RIGHTS_POLL_FD_READWRITE
+                    | wasi_fs::RIGHTS_FD_READ
+                    | wasi_fs::RIGHTS_FD_READDIR;
+                let res =
+                    unsafe { wasi_fs::path_open(3, 0, path.as_str(), oflag, right, 0, fdflag) };
```
Comment (L-jasmine) [inline src/internal_module/fs.rs:628]: the param `fd` in `path_xxx` is mean preopen dir fd.

if I preopen dir are `dir_a`  and `dir_b`. they `fd` are 3 and 4.
if I want open `dir_a/abc/text.txt`, it code is `wasi_fs::path_open(3,0,"abc/text.txt",oflag, right, 0, fdflag)`.
so the `fd` is not always `3`.

your can get the preopen dir by this way.(https://github.com/rust-lang/rust/blob/d8613f792c11d6d348b15eee79da561323fa0199/library/std/src/sys/wasi/fs.rs#L696)

maybe your need copy this function to `wasi_fs.rs`

**Comment by L-jasmine** [comment]
@Puellaquae  please run `cargo fmt`

**Comment by L-jasmine** [comment]
@Puellaquae  please clean [warning](https://github.com/second-state/wasmedge-quickjs/actions/runs/3210856008/jobs/5248609081#step:9:10).

**Comment by L-jasmine** [comment]
@Puellaquae 
Why it run 6h+?
https://github.com/second-state/wasmedge-quickjs/actions/runs/3214392892/jobs/5286525322

File: .cargo/config.toml [hunk 4]
```diff
@@ -1,3 +1,6 @@
 [build]
 target="wasm32-wasi"
-rustflags="-C target-feature=+bulk-memory"
\ No newline at end of file
+rustflags="-C target-feature=+bulk-memory"
+
+[target.wasm32-wasi]
+runner="timeout 60s wasmedge --dir=.:. "
```
Comment (L-jasmine) [inline .cargo/config.toml:6]: Can it return a error when timeout?
If some bug let test timeout, CI will get this error?

Comment (Puellaquae) [reply .cargo/config.toml:6]: I not sure. In my local, it won't timeout. If timeout, it just be killed without throwing any error.

Comment (L-jasmine) [reply .cargo/config.toml:6]: I think need delete `timeout 60s`. because it will hide some bug.

Comment (hydai) [reply .cargo/config.toml:6]: FYI, use the workflow timeout instead: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepstimeout-minutes

Comment (hydai) [reply .cargo/config.toml:6]: You can set the expected timeout of the current step. If it really takes so long, then letting it fail is fine.

Comment (Puellaquae) [reply .cargo/config.toml:6]: If some test blocked for some reason, it's need 6h to wait github action to kill it. `timeout 60s` is just make it fail faster.

Comment (Puellaquae) [reply .cargo/config.toml:6]: I will change it to github action's timeout.

**Comment by Puellaquae** [comment]
@L-jasmine poll fd write will cause a infinite loop in ubuntu, while in WSL1 it won't happen. 

```rust
fn main() -> Result<(), wasi::Errno> {
    unsafe {
        let fd = wasi::path_open(
            3,
            0,
            "test.txt",
            0,
            wasi::RIGHTS_FD_WRITE | wasi::RIGHTS_POLL_FD_READWRITE,
            0,
            wasi::FDFLAGS_NONBLOCK,
        )?;
        println!("{}", fd);
        let subs = vec![wasi::Subscription {
            userdata: 0,
            u: wasi::SubscriptionU {
                tag: 2,
                u: wasi::SubscriptionUU {
                    fd_write: wasi::SubscriptionFdReadwrite {
                        file_descriptor: fd as u32,
                    },
                },
            },
        }];
        let mut events = vec![wasi::Event {
            userdata: 0,
            error: wasi::ERRNO_SUCCESS,
            type_: wasi::EVENTTYPE_FD_WRITE,
            fd_readwrite: wasi::EventFdReadwrite {
                nbytes: 0,
                flags: 0,
            },
        }];
        let n = wasi::poll_oneoff(subs.as_ptr(), events.as_mut_ptr(), subs.len())?;
        println!("{}", n);
    };
    Ok(())
}
```

**Comment by Puellaquae** [comment]
> > @L-jasmine poll fd write will cause a infinite loop in ubuntu, while in WSL1 it won't happen.
> > ```rust
> > fn main() -> Result<(), wasi::Errno> {
> >     unsafe {
> >         let fd = wasi::path_open(
> >             3,
> >             0,
> >             "test.txt",
> >             0,
> >             wasi::RIGHTS_FD_WRITE | wasi::RIGHTS_POLL_FD_READWRITE,
> >             0,
> >             wasi::FDFLAGS_NONBLOCK,
> >         )?;
> >         println!("{}", fd);
> >         let subs = vec![wasi::Subscription {
> >             userdata: 0,
> >             u: wasi::SubscriptionU {
> >                 tag: 2,
> >                 u: wasi::SubscriptionUU {
> >                     fd_write: wasi::SubscriptionFdReadwrite {
> >                         file_descriptor: fd as u32,
> >                     },
> >                 },
> >             },
> >         }];
> >         let mut events = vec![wasi::Event {
> >             userdata: 0,
> >             error: wasi::ERRNO_SUCCESS,
> >             type_: wasi::EVENTTYPE_FD_WRITE,
> >             fd_readwrite: wasi::EventFdReadwrite {
> >                 nbytes: 0,
> >                 flags: 0,
> >             },
> >         }];
> >         let n = wasi::poll_oneoff(subs.as_ptr(), events.as_mut_ptr(), subs.len())?;
> >         println!("{}", n);
> >     };
> >     Ok(())
> > }
> > ```
> 
> @hydai

poll fd read also has same problem.

**Comment by Puellaquae** [comment]
`path_readlink` is not work well
```rust
fn main() -> Result<(), wasi::Errno> {
    unsafe {
        wasi::path_symlink("./test.txt", 3, "./test-link.txt")?;
        let stat = wasi::path_filestat_get(3, 0, "./test-link.txt")?;
        println!("{:?}", stat);
        let mut buf = [0;1024];
        let len = wasi::path_readlink(3, "./test-link.txt", buf.as_mut_ptr(), buf.len())?;
        println!("{:?}", String::from_utf8_lossy(&buf[0..len]));
    };
    Ok(())
}
```
It will return error: `Error: Errno { code: 28, name: "INVAL", message: "Invalid argument." }`
If the link file is under a folder like:
```rust
fn main() -> Result<(), wasi::Errno> {
    unsafe {
        wasi::path_symlink("./test.txt", 3, "./test/test-link.txt")?;
        let stat = wasi::path_filestat_get(3, 0, "./test/test-link.txt")?;
        println!("{:?}", stat);
        let mut buf = [0;1024];
        let len = wasi::path_readlink(3, "./test/test-link.txt", buf.as_mut_ptr(), buf.len())?;
        println!("{:?}", String::from_utf8_lossy(&buf[0..len]));
    };
    Ok(())
}
```
It will return error: `Error: Errno { code: 44, name: "NOENT", message: "No such file or directory." }`
@hydai

**Comment by hydai** [comment]
We are focusing on some high-priority issues. If you encounter any bugs, please raise issues on the WasmEdge project with the detailed reproduction steps and the environment information. Thanks.

**Comment by L-jasmine** [comment]
> `path_readlink` is not work well
> 
> ```rust
> fn main() -> Result<(), wasi::Errno> {
>     unsafe {
>         wasi::path_symlink("./test.txt", 3, "./test-link.txt")?;
>         let stat = wasi::path_filestat_get(3, 0, "./test-link.txt")?;
>         println!("{:?}", stat);
>         let mut buf = [0;1024];
>         let len = wasi::path_readlink(3, "./test-link.txt", buf.as_mut_ptr(), buf.len())?;
>         println!("{:?}", String::from_utf8_lossy(&buf[0..len]));
>     };
>     Ok(())
> }
> ```
> 
> It will return error: `Error: Errno { code: 28, name: "INVAL", message: "Invalid argument." }` If the link file is under a folder like:
> 
> ```rust
> fn main() -> Result<(), wasi::Errno> {
>     unsafe {
>         wasi::path_symlink("./test.txt", 3, "./test/test-link.txt")?;
>         let stat = wasi::path_filestat_get(3, 0, "./test/test-link.txt")?;
>         println!("{:?}", stat);
>         let mut buf = [0;1024];
>         let len = wasi::path_readlink(3, "./test/test-link.txt", buf.as_mut_ptr(), buf.len())?;
>         println!("{:?}", String::from_utf8_lossy(&buf[0..len]));
>     };
>     Ok(())
> }
> ```
> 
> It will return error: `Error: Errno { code: 44, name: "NOENT", message: "No such file or directory." }` @hydai

I think `symlink` is not very important.

**Comment by L-jasmine** [comment]
> > > @L-jasmine poll fd write will cause a infinite loop in ubuntu, while in WSL1 it won't happen.
> > > ```rust
> > > fn main() -> Result<(), wasi::Errno> {
> > >     unsafe {
> > >         let fd = wasi::path_open(
> > >             3,
> > >             0,
> > >             "test.txt",
> > >             0,
> > >             wasi::RIGHTS_FD_WRITE | wasi::RIGHTS_POLL_FD_READWRITE,
> > >             0,
> > >             wasi::FDFLAGS_NONBLOCK,
> > >         )?;
> > >         println!("{}", fd);
> > >         let subs = vec![wasi::Subscription {
> > >             userdata: 0,
> > >             u: wasi::SubscriptionU {
> > >                 tag: 2,
> > >                 u: wasi::SubscriptionUU {
> > >                     fd_write: wasi::SubscriptionFdReadwrite {
> > >                         file_descriptor: fd as u32,
> > >                     },
> > >                 },
> > >             },
> > >         }];
> > >         let mut events = vec![wasi::Event {
> > >             userdata: 0,
> > >             error: wasi::ERRNO_SUCCESS,
> > >             type_: wasi::EVENTTYPE_FD_WRITE,
> > >             fd_readwrite: wasi::EventFdReadwrite {
> > >                 nbytes: 0,
> > >                 flags: 0,
> > >             },
> > >         }];
> > >         let n = wasi::poll_oneoff(subs.as_ptr(), events.as_mut_ptr(), subs.len())?;
> > >         println!("{}", n);
> > >     };
> > >     Ok(())
> > > }
> > > ```
> > 
> > 
> > @hydai
> 
> poll fd read also has same problem.

in linux, epoll a normal file will get a `Operation not permitted`.
normal file read & write not need wait.
so, we can skip poll a file, just think it is ready to read or write.

**Comment by L-jasmine** [comment]
@Puellaquae  if you think this pr is time to merge, please @ me. Thinks.

**Comment by Puellaquae** [comment]
@L-jasmine  Most of fs api is implemented and passed nodejs's tests. `ReadStream` and `WriteStream` only has basic function. `chmod`, `chown` and `watch` is unsupported. And `symlink` on dir, `readlink` and `realpath` is not work well now.

**Review by L-jasmine — APPROVED** [review]

**Pull request merged**